### PR TITLE
[Mono][LLVM][ARM] Implemented Arm64 Abs intrinsic

### PR DESCRIPTION
**Pull request by naricc**
Implements AdvSimd.Abs intrinsic on arm64 with LLVM.

### CODE REVIEW

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -10501,6 +10517,24 @@ add_intrinsic (LLVMModuleRef module, int id)
...
+		break;
+	case INTRINS_ADV_SIMD_ABS_INT16:
+		intrins = add_intrins1 (module, id, sse_i2_t);
+		break;
+	case INTRINS_ADV_SIMD_ABS_INT32:
+		intrins = add_intrins1 (module, id, sse_i4_t);
+		break;
+	case INTRINS_ADV_SIMD_ABS_INT64:
+		intrins = add_intrins1 (module, id, sse_i8_t);
+		break;
```
Comment (EgorBo) [inline src/mono/mono/mini/mini-llvm.c:10537]: The whole function looks like the work to define signatures for intrinsics could be done via macros in `llvm-intrinsics.h` instead.

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -8229,6 +8229,8 @@ parse_cpu_features (const gchar *attr)
 		feature = MONO_CPU_ARM64_BASE;
 	else if (!strcmp (attr + prefix, "crc"))
 		feature = MONO_CPU_ARM64_CRC;
+	else if (!strcmp (attr + prefix, "simd"))
+		feature = MONO_CPU_ARM64_ADVSIMD;
```
[the remark this answers is not in the document]
Comment (imhameed) [reply src/mono/mono/mini/aot-compiler.c:8233]: My personal (weakly-held) preference would be to reuse clang/gcc-style feature names for this. They both use "simd".

File: src/mono/mono/mini/mini-llvm.c (hunk 1 above)
Comment (naricc) [reply src/mono/mono/mini/mini-llvm.c:10600]: Do we have macros to do this type of thing, or were you suggesting I create some? I would have think think a little about how to do it; this function predates my changes and would need some reworking.

Comment (EgorBo) [reply src/mono/mono/mini/mini-llvm.c:10600]: Yeah, I mean in general, if you want to eventually implement the full set of intrinsics it makes sense to minimize verbosity of the changes.

**Review by EgorBo — APPROVED** [review]

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 3]
```diff
@@ -941,6 +945,41 @@ emit_arm64_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatur
 		if (info->op != 0)
 			return emit_simd_ins_for_sig (cfg, klass, info->op, info->instc0, arg0_type, fsig, args);
 	}
+	
+	if (is_hw_intrinsics_class (klass, "AdvSimd", &is_64bit)) {
+		info = lookup_intrins_info (advsimd_methods, sizeof(advsimd_methods), cmethod);
```
Comment (CoffeeFlux) [inline src/mono/mono/mini/simd-intrinsics-netcore.c:950]:
```suggestion
		info = lookup_intrins_info (advsimd_methods, sizeof (advsimd_methods), cmethod);	
```

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 4]
```diff
@@ -941,6 +945,41 @@ emit_arm64_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatur
...
+	
+	if (is_hw_intrinsics_class (klass, "AdvSimd", &is_64bit)) {
+		info = lookup_intrins_info (advsimd_methods, sizeof(advsimd_methods), cmethod);	
+
+		if (!info)
+			return NULL;
+
+		supported = (mini_get_cpu_features (cfg) & MONO_CPU_ARM64_ADVSIMD) != 0;
+
+		switch (info -> id ) {
```
Comment (CoffeeFlux) [inline src/mono/mono/mini/simd-intrinsics-netcore.c:957]:
```suggestion
		switch (info -> id) {
```

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 5]
```diff
@@ -941,6 +945,41 @@ emit_arm64_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatur
...
+			case SN_Abs: {
+				SimdOp op = (SimdOp)0;
+				switch(get_underlying_type(fsig->params[0])){
```
Comment (CoffeeFlux) [inline src/mono/mono/mini/simd-intrinsics-netcore.c:960]:
```suggestion
				switch (get_underlying_type (fsig->params [0])) {
```

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 6]
```diff
@@ -941,6 +945,41 @@ emit_arm64_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatur
...
+				case MONO_TYPE_I2:
+					op = SIMD_OP_LLVM_I16ABS;
+				case MONO_TYPE_I4:
+					op = SIMD_OP_LLVM_I32ABS;
+				case MONO_TYPE_I8:
+					op = SIMD_OP_LLVM_I64ABS;
+					break;
+				}
+
+				return emit_simd_ins_for_sig(cfg, klass, OP_XOP_X_X, op, arg0_type, fsig, args);
```
Comment (CoffeeFlux) [inline src/mono/mono/mini/simd-intrinsics-netcore.c:977]:
```suggestion
				return emit_simd_ins_for_sig (cfg, klass, OP_XOP_X_X, op, arg0_type, fsig, args);
```

File: src/mono/mono/mini/llvm-intrinsics.h [hunk 7]
```diff
@@ -263,6 +263,12 @@ INTRINS(AARCH64_SHA256SU0, aarch64_crypto_sha256su0)
 INTRINS(AARCH64_SHA256SU1, aarch64_crypto_sha256su1)
 INTRINS(AARCH64_SHA256H, aarch64_crypto_sha256h)
 INTRINS(AARCH64_SHA256H2, aarch64_crypto_sha256h2)
+INTRINS_OVR(ADV_SIMD_ABS_FLOAT, fabs)
+INTRINS_OVR(ADV_SIMD_ABS_DOUBLE, fabs)
```
Comment (vargaz) [inline src/mono/mono/mini/llvm-intrinsics.h:267]: These should have a AARCH64 prefix ?

Comment (naricc) [reply src/mono/mono/mini/llvm-intrinsics.h:267]: That seems reasonable.

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 8]
```diff
@@ -941,6 +945,41 @@ emit_arm64_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatur
...
+				case MONO_TYPE_I4:
+					op = SIMD_OP_LLVM_I32ABS;
+				case MONO_TYPE_I8:
+					op = SIMD_OP_LLVM_I64ABS;
+					break;
+				}
+
+				return emit_simd_ins_for_sig (cfg, klass, OP_XOP_X_X, op, arg0_type, fsig, args);
+			}
+		}
```
Comment (imhameed) [inline src/mono/mono/mini/simd-intrinsics-netcore.c:979]:
```suggestion
		switch (info -> id) {
		case SN_Abs: {
			SimdOp op = (SimdOp)0;
			switch (get_underlying_type (fsig->params [0])) {
			case MONO_TYPE_R8:
				op = SIMD_OP_LLVM_DABS;
			case MONO_TYPE_R4:
				op = SIMD_OP_LLVM_FABS;
				break;
			case MONO_TYPE_I1:
				op = SIMD_OP_LLVM_I8ABS;
			case MONO_TYPE_I2:
				op = SIMD_OP_LLVM_I16ABS;
			case MONO_TYPE_I4:
				op = SIMD_OP_LLVM_I32ABS;
			case MONO_TYPE_I8:
				op = SIMD_OP_LLVM_I64ABS;
				break;
			}

			return emit_simd_ins_for_sig (cfg, klass, OP_XOP_X_X, op, arg0_type, fsig, args);
		}
		}
```

**Review by imhameed — APPROVED** [review]

**Review by kunalspathak — APPROVED** [review]
LGTM. So this is the first `AdvSimd.Arm64` API that is being supported in mono?

**Comment by naricc** [comment]
@kunalspathak 

> LGTM. So this is the first `AdvSimd.Arm64` API that is being supported in mono?

Yes. I am working on implement more; the issue listing them all is here: https://github.com/dotnet/runtime/issues/42266

**Comment by kunalspathak** [comment]
> Yes. I am working on implement more; the issue listing them all is here: #42266

Got it. I have added a reference to the documentation that I worked few months back to explain what every AdvSimd.Arm64 intrinsic API does. In case it is helpful to you or anyone working on this feature. 

https://kunalspathak.github.io/tags/#intrinsics

**Pull request merged**